src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxDocumentConverter.cs

```diff
@@ -0,0 +1,287 @@
+                        return;
+                    }
+                    @long = (long)lazyNumber;
+                    @double = lazyNumber.ToDouble(CultureInfo.InvariantCulture);
+
+                    scope.Write(field.Id, lazyNumber.Inner.AsSpan(), @long, @double, ref entryWriter);
+                    return;
+
+                case ValueType.String:
+                    scope.Write(field.Id, (string)value, ref entryWriter);
+                    return;
+
+                case ValueType.LazyCompressedString:
+                case ValueType.LazyString:
+                    LazyStringValue lazyStringValue;
+                    if (valueType == ValueType.LazyCompressedString)
+                        lazyStringValue = ((LazyCompressedStringValue)value).ToLazyStringValue();
+                    else
+                        lazyStringValue = (LazyStringValue)value;
+                    scope.Write(field.Id, lazyStringValue.AsSpan(), ref entryWriter);
+                    return;
+
+                case ValueType.Enum:
+                    scope.Write(field.Id, value.ToString(), ref entryWriter);
+                    return;
+
+                case ValueType.Boolean:
+                    scope.Write(field.Id, (bool)value ? _trueLiteral : _falseLiteral, ref entryWriter);
+                    return;
+
+                case ValueType.DateTime:
+                    var dateTime = (DateTime)value;
+                    var dateAsBytes = dateTime.GetDefaultRavenFormat();
+                    scope.Write(field.Id, dateAsBytes, ref entryWriter);
+                    return;
+
+                case ValueType.DateTimeOffset:
+                    var dateTimeOffset = (DateTimeOffset)value;
+                    var dateTimeOffsetBytes = dateTimeOffset.UtcDateTime.GetDefaultRavenFormat(isUtc: true);
+                    scope.Write(field.Id, dateTimeOffsetBytes, ref entryWriter);
+                    return;
+
+                case ValueType.TimeSpan:
+                    var timeSpan = (TimeSpan)value;
+                    using (_allocator.Allocate(256 ,out var buffer))
+                    {
+                        if (Utf8Formatter.TryFormat(timeSpan, buffer.ToSpan(), out var bytesWritten, _timeSpanFormat) == false)
+                            throw new Exception($"Cannot convert {field.Name} as double into bytes.");
+                        buffer.Truncate(bytesWritten);
+                        scope.Write(field.Id, buffer.ToSpan(), ref entryWriter);
+                    }
+                    return;
+
+                case ValueType.Convertible:
+                    scope.Write(field.Id, ((IConvertible)value).ToString(CultureInfo.InvariantCulture), ref entryWriter);
+                    return;
+
+                case ValueType.Enumerable:
+                    var iterator = (IEnumerable)value;
+                    var enumerableScope = new EnumerableWriterScope(field.Id, ref entryWriter, scope.GetLengthList(), _allocator);
```

Comment:
better to have construction like:
```
using (var enumerableScope = new EnumerableWriterScope(field.Id, ref entryWriter, scope.GetLengthList(), _allocator))
{
}
```

Reply:
Agreed, but I need to have a reference entryWriter in Dispose method and I can't save it as a prop.